src/allocation_engine/interval_tree.rs

```diff
@@ -467,142 +499,178 @@ mod tests {
+        assert_eq!(tree.search(&Range::new(0x111, 0x1fe).unwrap()), None);
     }
 
     #[test]
     fn test_search_superset() {
-        let left_child = InnerNode::new(Range::new(0x90, 0x99).unwrap(), NodeState::Free);
-        let right_child = InnerNode::new(Range::new(0x200, 0x2ff).unwrap(), NodeState::Free);
-        let mut root_node = InnerNode::new(Range::new(0x100, 0x110).unwrap(), NodeState::Allocated);
-        root_node.left = Some(Box::new(left_child));
-        root_node.right = Some(Box::new(right_child.clone()));
+        let mut tree = Box::new(InnerNode::new(
+            Range::new(0x100, 0x110).unwrap(),
+            NodeState::Allocated,
+        ));
+        let right_child = InnerNode::new(Range::new(0x200, 0x2FF).unwrap(), NodeState::Free);
+        let left_child = InnerNode::new(Range::new(0x90, 0x9F).unwrap(), NodeState::Free);
+
+        tree = tree.insert(left_child.key, left_child.node_state).unwrap();
+        tree = tree
+            .insert(right_child.key, right_child.node_state)
+            .unwrap();
 
         assert_eq!(
-            root_node.search_superset(&Range::new(0x100, 0x100).unwrap()),
-            Some(&root_node)
+            tree.search_superset(&Range::new(0x100, 0x100).unwrap()),
+            Some(&(*tree))
         );
         assert_eq!(
-            root_node.search_superset(&Range::new(0x200, 0x201).unwrap()),
+            tree.search_superset(&Range::new(0x90, 0x95).unwrap()),
+            Some(&left_child)
+        );
+        assert_eq!(
+            tree.search_superset(&Range::new(0x200, 0x201).unwrap()),
             Some(&right_child)
         );
         assert_eq!(
-            root_node.search_superset(&Range::new(0x200, 0x2ff).unwrap()),
+            tree.search_superset(&Range::new(0x200, 0x2FF).unwrap()),
             Some(&right_child)
         );
         assert_eq!(
-            root_node.search_superset(&Range::new(0x210, 0x210).unwrap()),
+            tree.search_superset(&Range::new(0x210, 0x210).unwrap()),
             Some(&right_child)
         );
         assert_eq!(
-            root_node.search_superset(&Range::new(0x2ff, 0x2ff).unwrap()),
+            tree.search_superset(&Range::new(0x2FF, 0x2FF).unwrap()),
             Some(&right_child)
         );
         assert_eq!(
-            root_node.search_superset(&Range::new(0x2ff, 0x300).unwrap()),
+            tree.search_superset(&Range::new(0x2FF, 0x300).unwrap()),
             None
         );
         assert_eq!(
-            root_node.search_superset(&Range::new(0x300, 0x300).unwrap()),
+            tree.search_superset(&Range::new(0x300, 0x300).unwrap()),
             None
         );
         assert_eq!(
-            root_node.search_superset(&Range::new(0x1ff, 0x300).unwrap()),
+            tree.search_superset(&Range::new(0x1ff, 0x300).unwrap()),
             None
         );
     }
 
+    fn is_balanced(tree: Option<Box<InnerNode>>) -> bool {
```

Comment:
How is this going to be used? It would be better to have this as a function of `InnerNode` (i.e. receiving a reference to `&self` as a parameter). It's also a bit strange that the function receives an `Option` as a parameter because for `None` we shouldn't be asking the question about the tree being balanced or not. I guess this was added such that we can use recursion, but the problem can be solved with loops as well.